### PR TITLE
Improve blame file type detection (#1803)

### LINKED ISSUE #1803: 🐛 `git blame -C` loses syntax highlighting
The [`-C` flag](https://git-scm.com/docs/git-blame#Documentation/git-blame.txt--Cltnumgt) for `git blame` is useful for tracking blame beyond code that was moved from one file to another. Unfortunately, it seems to cause delta to lose its syntax highlighting - the blame still has grouped background highlights in different shades of gray, but the code itself isn't highlighted. Here is an asciinema recording demonstrating the issue by running `git blame src/handlers/blame.rs` and `git blame -C src/handlers/blame.rs` in a clone of this repository (at commit a01141b72001f4c630d77cf5274267d7638851e4):

[asciicast](https://asciinema.org/a/IYaYidJbWmM7wwXDGrtX6jRmo)

### PR DESCRIPTION
**Pull request by dvermd**
closes #1803

### CODE REVIEW

File: src/utils/process.rs [hunk 1]
```diff
@@ -726,6 +726,9 @@ pub mod tests {
         let args = make_string_vec(&["git", "-c", "a=b", "blame", "main.rs"]);
         assert_eq!(guess_git_blame_filename(&args), Args("main.rs".into()));
 
+        let args = make_string_vec(&["git", "blame", "-w", "-C", "main.rs"]);
+        assert_eq!(guess_git_blame_filename(&args), Args("main.rs".into()));
+
```
Comment (th1000s) [inline src/utils/process.rs:731]: That unfortunately doesn't really fix it: Now `-w` takes an argument (when it does not), which means it "eats" the following `-C`. The problem is `-C[<num>]`, which only takes a non-whitespace argument (or none) -- but this not really modeled at the moment.

Removing `-C` from the previous list should work for most cases. In the future a fallback to take the last argument containing a `.` should be added. And make it use the `CallingProcess` enum as mentioned in the TODO there.

Comment (dvermd) [reply src/utils/process.rs:731]: I've been tricked by only looking at the synopsis part of the git blame man page when adding `-w` to this list.

I'll look at addressing this TODO and making a variant in `CallingProcess`.

File: src/utils/process.rs [hunk 2]
```diff
@@ -702,53 +653,6 @@ pub mod tests {
...
-            "--date",
-            "now",
-        ]);
-        assert_eq!(guess_git_blame_filename(&args), Args("hello.txt".into()));
-
-        let args = make_string_vec(&["git", "blame", "-s", "-f", "--", "hello.txt"]);
-        assert_eq!(guess_git_blame_filename(&args), Args("hello.txt".into()));
-
-        let args = make_string_vec(&["git", "blame", "--", "--not.an.argument"]);
-        assert_eq!(
```
Comment (th1000s) [inline src/utils/process.rs:731]: Is this case still tested? And the failure cases, such as `git blame --help.txt`?

**Review by th1000s — COMMENTED** [review]
Nice, a few small things.

File: src/utils/process.rs (hunk 2 above)
Comment (dvermd) [reply src/utils/process.rs:731]: I added more tests

File: src/utils/process.rs [hunk 3]
```diff
@@ -925,67 +856,116 @@ pub mod tests {
...
-            Some("once".into())
+            calling_process_cmdline(ProcInfo::new(), describe_calling_process),
+            Some(CallingProcess::GitBlame(CommandLine {
+                long_options: [].into(),
+                short_options: [].into(),
+                last_arg: Some("once".into())
+            }))
         );
-        calling_process_cmdline(ProcInfo::new(), guess_git_blame_filename);
+        // ignored: dropping causes a panic while panicking, so can't test
```
[the remark this answers is not in the document]
Comment (dvermd) [reply src/utils/process.rs]: Indeed, removed

File: src/handlers/blame.rs [hunk 4]
```diff
@@ -77,8 +77,7 @@ impl<'a> StateMachine<'a> {
 
                 // Emit syntax-highlighted code
                 if matches!(self.state, State::Unknown) {
```
[the remark this answers is not in the document]
Comment (dvermd) [reply src/handlers/blame.rs]: fixed

File: src/utils/process.rs [hunk 5]
```diff
@@ -223,49 +204,19 @@ fn is_git_binary(git: &str) -> bool {
...
-    }
-}
-
 // Given `--aa val -bc -d val e f -- ...` return
 // ({"--aa"}, {"-b", "-c", "-d"})
 fn parse_command_line<'a>(args: impl Iterator<Item = &'a str>) -> CommandLine {
     let mut long_options = HashSet::new();
     let mut short_options = HashSet::new();
     let mut last_arg = None;
+    let mut in_pos_args = false;
```
[the remark this answers is not in the document]
Comment (dvermd) [reply src/utils/process.rs]: I used your naming

File: src/utils/process.rs [hunk 6]
```diff
@@ -26,7 +27,9 @@ impl CallingProcess {
             CallingProcess::GitDiff(cmd) if cmd.long_options.contains("--relative") => true,
             CallingProcess::GitShow(cmd, _) if cmd.long_options.contains("--relative") => true,
             CallingProcess::GitLog(cmd) if cmd.long_options.contains("--relative") => true,
-            CallingProcess::GitGrep(_) | CallingProcess::OtherGrep => true,
+            CallingProcess::GitGrep(_)
+            | CallingProcess::GitBlame(_)
+            | CallingProcess::OtherGrep => true,
```
[the remark this answers is not in the document]
Comment (dvermd) [reply src/utils/process.rs]: They're now sorted

**Comment by th1000s** [comment]
Looks good, thank you!

**Pull request merged**